File: Problems/Algorithms/234. Palindrome Linked List/palindrome_linked_list.py
#!/usr/bin/env python3

from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow, fast = head, head
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next

        prev, slow, prev.next = slow, slow.next, None
        while slow:
            slow.next, prev, slow = prev, slow, slow.next

        start, end = head, prev
        while end:
            if start.val != end.val:
                return False
            start, end = start.next, end.next
        return True

File: Problems/Algorithms/234. Palindrome Linked List/palindrome_linked_list.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn is_palindrome(mut head: Option<Box<ListNode>>) -> bool {
        let mut values = vec![];
        while let Some(node) = head {
            values.push(node.val);
            head = node.next;
        }

        let reversed = values.iter().rev().cloned().collect::<Vec<_>>();
        values == reversed
    }
}
